File: frontend/src/types/listingDraft.ts
import { ISODate } from '@/utils/date';

import type { Listing } from './listing';

export const LISTING_DRAFT_STATUS_LIST = [
  'pending',
  'unique',
  'duplicate_url',
  'duplicate_content',
  'error',
] as const;

export type ListingDraftStatus = (typeof LISTING_DRAFT_STATUS_LIST)[number];

export type GroundedItem<T = string> = {
  value: T;
  quote: string | null;
};

export type ListingExtraction = {
  title: string;
  company: string;
  domain: string;
  location: string | null;
  description: string;
  postedDate: ISODate | null;
  skills: GroundedItem<string>[];
  requirements: GroundedItem<string>[];
};

type BaseListingDraft = {
  id: string;
  url: string;
};

export type ListingDraftUnique = BaseListingDraft & {
  status: 'unique';
  listing: ListingExtraction;
  html: string | null;
};

export type ListingDraftDuplicateUrl = BaseListingDraft & {
  status: 'duplicate_url';
  duplicateOf: Listing;
};

export type ListingDraftDuplicateContent = BaseListingDraft & {
  status: 'duplicate_content';
  listing: ListingExtraction;
  duplicateOf: Listing;
  html: string | null;
};

export type ListingDraftError = BaseListingDraft & {
  status: 'error';
  error: string;
  html: string | null;
};

export type ListingDraftPending = BaseListingDraft & {
  status: 'pending';
};

export type ListingDraft =
  | ListingDraftPending
  | ListingDraftUnique
  | ListingDraftDuplicateUrl
  | ListingDraftDuplicateContent
  | ListingDraftError;

File: backend/app/schemas/listing_draft.py
from enum import StrEnum
from typing import Annotated, Literal
from uuid import UUID, uuid4

from pydantic import Field, HttpUrl

from app.schemas.listing import Listing, ListingBase
from shared.schemas.types import CamelModel


class DraftStatusEnum(StrEnum):
  UNIQUE = 'unique'
  DUPLICATE_URL = 'duplicate_url'
  DUPLICATE_CONTENT = 'duplicate_content'
  ERROR = 'error'


class ListingExtraction(ListingBase):
  description: str
  skills: list[str] = Field(default_factory=list)
  requirements: list[str] = Field(default_factory=list)


class BaseListingDraft(CamelModel):
  id: UUID = Field(default_factory=uuid4)
  url: HttpUrl


class ListingDraftUnique(BaseListingDraft):
  status: Literal[DraftStatusEnum.UNIQUE] = DraftStatusEnum.UNIQUE
  listing: ListingExtraction
  screenshot: str | None = None


class ListingDraftDuplicateUrl(BaseListingDraft):
  status: Literal[DraftStatusEnum.DUPLICATE_URL] = DraftStatusEnum.DUPLICATE_URL
  duplicate_of: Listing


class ListingDraftDuplicateContent(BaseListingDraft):
  status: Literal[DraftStatusEnum.DUPLICATE_CONTENT] = DraftStatusEnum.DUPLICATE_CONTENT
  listing: ListingExtraction
  duplicate_of: Listing
  screenshot: str | None = None


class ListingDraftError(BaseListingDraft):
  status: Literal[DraftStatusEnum.ERROR] = DraftStatusEnum.ERROR
  error: str
  screenshot: str | None = None


ListingDraft = Annotated[
  ListingDraftUnique | ListingDraftDuplicateUrl | ListingDraftDuplicateContent | ListingDraftError,
  Field(discriminator='status'),
]
